### PR TITLE
SCP-4441, SCP-4283 - apply inputs and create contract constraints

### PR DESCRIPTION
**Pull request by paluh**
Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [ ] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

File: marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/BuildConstraints.hs [hunk 1]
```diff
@@ -35,7 +84,137 @@ buildApplyInputsConstraints
                    -- in the contract.
   -> Redeemer v -- ^ The inputs to apply to the contract.
   -> Either ApplyInputsError (TxConstraints v)
-buildApplyInputsConstraints = error "not implemented"
```
Comment (paluh) [inline marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/BuildConstraints.hs:38]: @jhbertra should I make this function effectful so I can grab the current time if it is necessary to define lower time range boundry?

File: marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/BuildConstraints.hs [hunk 2]
```diff
@@ -35,7 +84,137 @@ buildApplyInputsConstraints
...
+    slotStart = error "not implemented"
+
+    marloweInputContent (V1.NormalInput c) = c
+    marloweInputContent (V1.MerkleizedInput c _ _) = c
+
+    marloweInputParty = marloweInputContent >>> \case
+      V1.IDeposit _ party _ _         -> Just party
+      V1.IChoice (V1.ChoiceId _ party) _ -> Just party
+      V1.INotify                      -> Nothing
+
```
Comment (paluh) [inline marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/BuildConstraints.hs:187]: @jhbertra The below pieces are literal copies of the code from the test suite. Do you think that we should extract these conversions into a separate place? Any suggestions where to put them in such a case?

File: marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/Constraints.hs [hunk 3]
```diff
@@ -46,6 +46,7 @@ data RoleTokenConstraints
   | SpendRoleTokens (Set Chain.AssetId)
   deriving (Eq, Ord, Show)
```
Comment (paluh) [inline marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/Constraints.hs:48]: I'm going to fix this accidental new line ;-)

File: marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/Server.hs [hunk 4]
```diff
@@ -217,12 +226,12 @@ execCreate solveConstraints loadWalletContext networkId era mStakeCredential ver
         ScriptDataInBabbageEra -> \(TxOut address _ _ _) -> address == AddressInEra (ShelleyAddressInEra ShelleyBasedEraBabbage) scriptAddress
 
 execApplyInputs
-  :: SlotConfig
```
Comment (paluh) [inline marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/Server.hs:220]: I see now that `IsMarloweVersion` is not really useful downstream. @jhbertra Should I drop this experiment or I can leave it as it is?

File: marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/BuildConstraints.hs [hunk 5]
```diff
@@ -35,7 +84,137 @@ buildApplyInputsConstraints
...
+  -- Construct inputs constraints.
+  -- Consume UTXOs containing Marlowe script.
+  tell $ mustConsumeMarloweOutput invalidBefore' invalidHereafter' redeemer
+
+  -- Consume UTXOs containing all necessary role tokens and send them back.
+  for_ requiredParties $ traverse_ $ \case
+    V1.Role role -> tell $ mustSpendRoleToken $ roleAssetId role
+    _ -> pure ()
+
+  -- Require signature of an every party which is autorhized through an address.
```
Comment (paluh) [inline marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/BuildConstraints.hs:125]: I'm going to fix the above typos.

File: marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/BuildConstraints.hs (hunk 1 above)
Comment (jhbertra) [reply marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/BuildConstraints.hs:38]: Yes, that sounds like a good idea.

File: marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/Api.hs [hunk 6]
```diff
@@ -751,41 +758,49 @@ instance Query.IsQuery ChainSyncQuery where
     GetProtocolParameters -> mempty
     GetSystemStart -> mempty
     GetEraHistory -> mempty
+    GetUTxOs utxos -> put . fmap unAddress . Set.toList $ utxos
```
Comment (jhbertra) [inline marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/Api.hs:761]: `fmap unAddress is no needed - `Address` has a `Binary` instance:

```suggestion
    GetUTxOs utxos -> put . Set.toList $ utxos
```

File: marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/Api.hs [hunk 7]
```diff
@@ -751,41 +758,49 @@ instance Query.IsQuery ChainSyncQuery where
...
   getQuery = \case
     TagGetSlotConfig        -> pure GetSlotConfig
     TagGetSecurityParameter -> pure GetSecurityParameter
     TagGetNetworkId -> pure GetNetworkId
     TagGetProtocolParameters -> pure GetProtocolParameters
     TagGetSystemStart -> pure GetSystemStart
     TagGetEraHistory -> pure GetEraHistory
+    TagGetUTxOs -> do
+      utxos <- get
+      pure . GetUTxOs . Set.fromList . fmap Address $ utxos
```
Comment (jhbertra) [inline marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/Api.hs:771]:
```suggestion
    TagGetUTxOs -> GetUTxOs . Set.fromList <$> get
```

File: marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/Database/PostgreSQL.hs [hunk 8]
```diff
@@ -555,6 +559,48 @@ queryTxOuts slotNo txId = HT.statement (Api.unTxId txId, fromIntegral slotNo) $
...
+             , asset.name :: bytea?
+             , assetOut.quantity :: bigint?
+          FROM chain.txOut         AS txOut
+          LEFT JOIN chain.assetOut AS assetOut ON assetOut.txOutId = txOut.txId AND assetOut.txOutIx = txOut.txIx
+          LEFT JOIN chain.asset    AS asset    ON asset.id = assetOut.assetId
+         WHERE txOut.address = ANY($1 :: bytea[])
+         ORDER BY txIx
+      |] (Fold foldRow mempty id)
+    where
+      -- foldRow :: IntMap Api.TransactionOutput -> ReadTxOutRow -> IntMap Api.TransactionOutput
```
Comment (jhbertra) [inline marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/Database/PostgreSQL.hs:584]: Why comment out the type signature?

File: marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/BuildConstraints.hs [hunk 9]
```diff
@@ -35,7 +91,137 @@ buildApplyInputsConstraints
...
+
+  -- Construct outputs constraints.
+  -- Require Marlowe output if the contract is not closed.
+  case possibleContinuation of
+    Just (state'@V1.State { accounts }, contract') -> do
+      let
+        datum' = V1.MarloweData params state' contract'
+        assets = moneyToAssets $ V1.totalBalance accounts
+      tell $ mustSendMarloweOutput assets datum'
+    Nothing -> pure ()
```
Comment (jhbertra) [inline marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/BuildConstraints.hs:159]:
```suggestion
for_ possibleContinuation \(state'@V1.State { accounts }, contract') -> do
    let
      datum' = V1.MarloweData params state' contract'
      assets = moneyToAssets $ V1.totalBalance accounts
    tell $ mustSendMarloweOutput assets datum'
```

File: marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/BuildConstraints.hs [hunk 10]
```diff
@@ -35,7 +91,137 @@ buildApplyInputsConstraints
...
+  -- For every payment require an output either to the role
+  -- payout script or to directly to the party address.
+  for payments \(V1.Payment _ payee token quantity) -> do
+    let
+      assets = case token of
+        V1.Token "" "" -> Assets (Lovelace $ fromInteger quantity) mempty
+        V1.Token cs tn -> do
+          let
+            assetId = toAssetId cs tn
+            quantity' = CS.Quantity $ fromInteger quantity
```
Comment (jhbertra) [inline marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/BuildConstraints.hs:170]:
```suggestion
            quantity' = fromIntegral quantity
```

File: marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/BuildConstraints.hs (hunk 2 above)
Comment (jhbertra) [reply marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/BuildConstraints.hs:194]: Yes, I do have a suggestion, though I would do this as a dedicated PR. We already have a `Language.Marlowe.Runtime.Cardano.Api` module which defines all the `Cardano.Api <-> Language.Marlowe.Runtime.ChainSync.Api`` conversions. Let's make a `Language.Marlowe.Runtime.Plutus.Api` module that does the same thing with `Plutus` types.

File: marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/Server.hs (hunk 4 above)
Comment (jhbertra) [reply marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/Server.hs:217]: Up to you - I've found that 90% of the time, passing a `MarloweVersion v` is more helpful than requiring a `IsMarloweVersion v` constraint. The big exception is in class instance headers, where `IsMarloweVersion v` can be very useful.

File: marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/Database/PostgreSQL.hs (hunk 8 above)
Comment (paluh) [reply marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/Database/PostgreSQL.hs:584]: `ReadTxOut` is a bit different in this case I include `TxId` - I'm going to probably drop the signature all together.

**Review by bwbush — APPROVED** [review]
Looks great.

**Review by dino-iog — APPROVED** [review]

**Pull request merged**